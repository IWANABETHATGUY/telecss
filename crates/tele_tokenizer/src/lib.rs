#![warn(missing_docs)]

//! CSS Tokenizer

use std::{iter::Peekable, str::CharIndices};

mod inner_state;
mod token;

use inner_state::*;
use tele_utils::*;
use token::{Pos, Token, TokenType};

#[derive(Debug)]
pub struct Tokenizer<'s> {
  source: &'s str,
  iter: Peekable<CharIndices<'s>>,
  line: usize,
  colnmu: usize,
  offset: usize,
  pre_cursor: Pos,
  cursor: Pos,
  current_state: State,
  buffer: Vec<char>,
  tokens: Vec<Token>,
}

impl<'s> From<&'s str> for Tokenizer<'s> {
  fn from(source: &'s str) -> Self {
    Tokenizer {
      source,
      iter: source.char_indices().peekable(),
      // position related
      line: 1,
      colnmu: 1,
      offset: 0,
      pre_cursor: Pos::default(),
      cursor: Pos::default(),
      // state machine related
      current_state: State::Initial,
      buffer: Vec::with_capacity(100),
      tokens: Vec::new(),
    }
  }
}

impl<'s> Tokenizer<'s> {
  pub fn tokenize(&mut self) -> &mut Vec<Token> {
    while let Some(&(offset, c)) = self.iter.peek() {
      self.current_state = match self.current_state {
        State::Initial => match c {
          c if is_whitespace(c) => State::WhiteSpace,
          '.' => State::MayBeNumber,
          _ => {
            self.advance(offset);
            State::Initial
          }
        },
        State::WhiteSpace => {
          if is_whitespace(c) {
            if is_newline(c) {
              self.line += 1;
              self.colnmu = 1;
            }
            self.consume(offset, c);
            State::WhiteSpace
          } else {
            self.emit(TokenType::WhiteSpace);
            State::Initial
          }
        }
        State::MayBeNumber => {
          self.consume(offset, c);
          if is_digit(c) {
            State::EOF
          } else {
            self.emit(TokenType::Delim);
            State::Initial
          }
        }
        State::EOF => break,
      };
    }

    &mut self.tokens
  }

  fn consume(&mut self, offset: usize, c: char) {
    // consume
    self.advance(offset);
    self.buffer.push(c);
  }

  fn advance(&mut self, offset: usize) {
    self.iter.next();
    self.colnmu += 1;
    self.offset = offset;
  }

  fn get_cursor(&self) -> Pos {
    Pos {
      offset: self.offset,
      line: self.line,
      column: self.colnmu,
    }
  }

  fn emit(&mut self, token_type: TokenType) {
    let cur_cursor = self.get_cursor();

    let mut token = Token::default();
    token.token_type = token_type;
    token.start_pos = self.pre_cursor;
    token.end_pos = cur_cursor;
    token.content.append(&mut self.buffer);

    self.pre_cursor = cur_cursor;
    self.tokens.push(token)
  }
}
