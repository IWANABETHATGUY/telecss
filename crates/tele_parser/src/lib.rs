#![warn(missing_docs)]
#![deny(unused_imports)]
#![deny(unused_variables)]

//! CSS Parser

mod error;
mod node;

pub use error::Result as ParserResult;
use error::{SyntaxError, SyntaxErrorKind};
pub use node::*;

use std::{cell::RefCell, iter::Peekable, slice::Iter};

use tele_tokenizer::*;

pub struct Parser<'s> {
  tokens: &'s Vec<Token<'s>>,
  iter: RefCell<Peekable<Iter<'s, Token<'s>>>>,
}

impl<'s> From<&'s Vec<Token<'s>>> for Parser<'s> {
  fn from(tokens: &'s Vec<Token<'s>>) -> Self {
    Self {
      tokens,
      iter: RefCell::new(tokens.iter().peekable()),
    }
  }
}

impl<'s> Parser<'s> {
  pub fn parse(&self) -> ParserResult<StyleSheetNode> {
    let mut ss_node = StyleSheetNode::default();

    while let Some(node) = self.parse_statements()? {
      ss_node.statements.push(node)
    }

    self.skip_ws();

    let next = self.peek();
    debug_assert!(next.is_eof());

    ss_node.loc.end = next.end_pos;

    Ok(ss_node)
  }

  fn peek(&self) -> &Token {
    let next = self.iter.borrow_mut().peek().map(|&token| token);
    // Because the last token will always be EOF2, and we will handle it reasonably.
    // So in general/theoretically, we don't encounter `None`.
    debug_assert!(next.is_some());
    unsafe { next.unwrap_unchecked() }
  }

  fn consume(&self) -> Option<&Token> {
    self.iter.borrow_mut().next()
  }

  fn parse_statements(&self) -> ParserResult<Option<StatementNode>> {
    self.skip_ws();

    let next = self.peek();

    if next.is_eof() {
      return Ok(None);
    }

    if next.is_at_keyword() {
      Ok(self.parse_at_rule().map(|node| StatementNode::AtRule(node)))
    } else {
      Ok(
        self
          .parse_rule_set()?
          .map(|node| StatementNode::RuleSet(node)),
      )
    }
  }

  fn parse_rule_set(&self) -> ParserResult<Option<RuleSetNode>> {
    let mut rule_set_node = RuleSetNode::default();

    let mut token = self.peek();
    rule_set_node.loc.start = token.start_pos;

    loop {
      token = self.peek();
      if token.is_eof() {
        self.asset_eof(token)?;
      } else if token.is_lcb() {
        self.consume();
        while let Some(node) = self.parse_decl()? {
          rule_set_node.declarations.push(node);
        }
      } else if token.is_rcb() {
        self.consume();
        break;
      } else {
        rule_set_node.prelude.push_str(&token.to_string());
        self.consume();
      }
    }

    rule_set_node.prelude = rule_set_node.prelude.trim_end().to_string();
    rule_set_node.loc.end = token.end_pos;

    Ok(Some(rule_set_node))
  }

  fn parse_at_rule(&self) -> Option<AtRuleNode> {
    unimplemented!()
  }

  fn parse_decl(&self) -> ParserResult<Option<DeclarationNode>> {
    self.skip_ws();

    // declaration name

    let next = self.peek();
    if next.is_rcb() {
      return Ok(None);
    }
    self.asset_eof(next)?;

    if !next.is_ident() {
      return Err(Box::new(SyntaxError::from(
        SyntaxErrorKind::InvalidDeclName,
      )));
    }

    let mut decl_node = DeclarationNode::default();
    decl_node.name = next.to_string();
    decl_node.loc.start = next.start_pos;

    self.consume();
    self.skip_ws();

    // :

    let next = self.peek();
    if !next.is_colon() {
      return Err(Box::new(SyntaxError::from(SyntaxErrorKind::MissingColon)));
    }
    self.asset_eof(next)?;
    self.consume();
    self.skip_ws();

    // declaration value

    let mut last_one: Option<&Token> = None;
    let mut second_last: Option<&Token> = None;
    loop {
      let token = self.peek();
      if token.is_eof() {
        self.asset_eof(token)?;
      } else if token.is_semi_colon() {
        self.consume();
        break;
      } else if token.is_rcb() {
        break;
      }

      self.consume();

      if !token.is_ws() {
        if last_one.is_some() {
          second_last = last_one;
        }
        last_one = Some(token);
      }

      decl_node.value.push_str(&token.to_string());
    }

    decl_node.important = second_last.is_some()
      && unsafe { second_last.unwrap_unchecked() }.is_delim(&[b'!'])
      && last_one.is_some()
      && unsafe { last_one.unwrap_unchecked() }.is_ident_with(b"important");

    last_one.map(|token| decl_node.loc.end = token.end_pos);

    Ok(Some(decl_node))
  }

  fn asset_eof(&self, token: &Token) -> ParserResult<()> {
    if token.is_eof() {
      return Err(Box::new(SyntaxError::from(SyntaxErrorKind::UnexpectedEOF)));
    }
    Ok(())
  }

  fn skip_ws(&self) {
    loop {
      let token = self.peek();
      if token.token_type == TokenType::WhiteSpace {
        self.consume();
      } else {
        break;
      }
    }
  }
}
