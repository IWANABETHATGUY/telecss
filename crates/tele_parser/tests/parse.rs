use insta::assert_debug_snapshot;
use tele_parser::Parser;
use tele_tokenizer::Tokenizer;

#[test]
fn ruleset_with_a_single_decl() {
  let mut tokenizer: Tokenizer = r"  .foo { color: red; }  ".into();
  let parser = Parser::from(tokenizer.tokenize().unwrap());
  assert_debug_snapshot!(parser.parse());
}

#[test]
fn ruleset_with_a_single_decl_no_trailing_semicolon() {
  let mut tokenizer: Tokenizer = r"  .foo { color: red }  ".into();
  let parser = Parser::from(tokenizer.tokenize().unwrap());
  assert_debug_snapshot!(parser.parse());
}

#[test]
fn ruleset_with_multi_decls() {
  let mut tokenizer: Tokenizer = r"
    .foo { color: red;
    background: url('bar.png')
  }  "
    .into();
  let parser = Parser::from(tokenizer.tokenize().unwrap());
  assert_debug_snapshot!(parser.parse());
}

#[test]
fn decl_with_important() {
  let mut tokenizer: Tokenizer = r"  .foo { color: red !important; }  ".into();
  let parser = Parser::from(tokenizer.tokenize().unwrap());
  assert_debug_snapshot!(parser.parse());
}

#[test]
fn skip_comments() {
  let mut tokenizer: Tokenizer = r"
    /* comment 1 */ .foo /* comment 2 */ {
    /* comment 3 */ color /* comment 4 */ :
    /* comment 5 */ red /* comment 6 */ !important;
    /* comment 7 */ } /* comment 8 */
  "
  .into();
  let parser = Parser::from(tokenizer.tokenize().unwrap());
  assert_debug_snapshot!(parser.parse());
}

#[test]
fn percentage_token() {
  let mut tokenizer: Tokenizer = r"
    .foo {
      height: 100%;
    }
  "
  .into();
  let parser = Parser::from(tokenizer.tokenize().unwrap());
  assert_debug_snapshot!(parser.parse());
}

#[test]
fn at_rule_without_block() {
  let mut tokenizer: Tokenizer = r#" @charset "urf-8"; "#.into();
  let parser = Parser::from(tokenizer.tokenize().unwrap());
  assert_debug_snapshot!(parser.parse());
}

#[test]
fn at_rule_with_block() {
  let mut tokenizer: Tokenizer = r#"
  @media screen and (min-width: 900px) {
    article {
      padding: 1rem 3rem;
    }
  }
  "#
  .into();
  let parser = Parser::from(tokenizer.tokenize().unwrap());
  assert_debug_snapshot!(parser.parse());
}

#[test]
fn nested_at_rule() {
  let mut tokenizer: Tokenizer = r#"
  @supports (display: flex) {
    @media screen and (min-width: 900px) {
      article {
        display: flex;
      }
    }
  }
  "#
  .into();
  let parser = Parser::from(tokenizer.tokenize().unwrap());
  assert_debug_snapshot!(parser.parse());
}
