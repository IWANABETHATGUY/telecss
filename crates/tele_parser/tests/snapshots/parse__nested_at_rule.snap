---
source: crates/tele_parser/tests/parse.rs
assertion_line: 96
expression: parser.parse()

---
Ok(
    StyleSheetNode {
        loc: Loc {
            start: Pos {
                offset: 0,
                line: 1,
                column: 1,
            },
            end: Pos {
                offset: 133,
                line: 9,
                column: 4,
            },
        },
        statements: [
            AtRule(
                AtRuleNode {
                    loc: Loc {
                        start: Pos {
                            offset: 4,
                            line: 2,
                            column: 5,
                        },
                        end: Pos {
                            offset: 30,
                            line: 2,
                            column: 31,
                        },
                    },
                    name: "supports",
                    prelude: "(display: flex)",
                    block: [
                        AtRule(
                            AtRuleNode {
                                loc: Loc {
                                    start: Pos {
                                        offset: 36,
                                        line: 3,
                                        column: 7,
                                    },
                                    end: Pos {
                                        offset: 73,
                                        line: 3,
                                        column: 44,
                                    },
                                },
                                name: "media",
                                prelude: "screen and (min-width: 900px)",
                                block: [
                                    RuleSet(
                                        RuleSetNode {
                                            loc: Loc {
                                                start: Pos {
                                                    offset: 80,
                                                    line: 4,
                                                    column: 8,
                                                },
                                                end: Pos {
                                                    offset: 120,
                                                    line: 6,
                                                    column: 9,
                                                },
                                            },
                                            prelude: "article",
                                            declarations: [
                                                DeclarationNode {
                                                    loc: Loc {
                                                        start: Pos {
                                                            offset: 98,
                                                            line: 5,
                                                            column: 10,
                                                        },
                                                        end: Pos {
                                                            offset: 111,
                                                            line: 5,
                                                            column: 23,
                                                        },
                                                    },
                                                    name: "display",
                                                    value: "flex",
                                                    important: false,
                                                },
                                            ],
                                        },
                                    ),
                                ],
                            },
                        ),
                    ],
                },
            ),
        ],
    },
)
